File: tools/benchmarks/bench_pipeline.py
# Benchmark: Data pipeline — filter, transform, aggregate
from dataclasses import dataclass
from typing import List

@dataclass
class Order:
    customer: str
    product:  str
    qty:      int
    price:    float
    shipped:  bool

orders: List[Order] = [
    Order("Alice",   "Widget A", 3, 9.99,  True),
    Order("Bob",     "Widget B", 1, 24.99, False),
    Order("Charlie", "Widget A", 5, 9.99,  True),
    Order("Diana",   "Widget C", 2, 49.99, True),
    Order("Eve",     "Widget B", 4, 24.99, True),
    Order("Frank",   "Widget A", 1, 9.99,  False),
    Order("Grace",   "Widget C", 3, 49.99, True),
]

revenue = sum(o.qty * o.price for o in orders if o.shipped)

top_customers = [
    o.customer
    for o in sorted(
        filter(lambda o: o.shipped, orders),
        key=lambda o: o.qty * o.price,
        reverse=True
    )
][:3]

print(f"Revenue: ${revenue}")
print(f"Top customers: {', '.join(top_customers)}")

File: tools/benchmarks/bench_pipeline.ts
// Benchmark: Data pipeline — filter, transform, aggregate
interface Order {
  customer: string;
  product:  string;
  qty:      number;
  price:    number;
  shipped:  boolean;
}

const orders: Order[] = [
  { customer: "Alice",   product: "Widget A", qty: 3, price: 9.99,  shipped: true  },
  { customer: "Bob",     product: "Widget B", qty: 1, price: 24.99, shipped: false },
  { customer: "Charlie", product: "Widget A", qty: 5, price: 9.99,  shipped: true  },
  { customer: "Diana",   product: "Widget C", qty: 2, price: 49.99, shipped: true  },
  { customer: "Eve",     product: "Widget B", qty: 4, price: 24.99, shipped: true  },
  { customer: "Frank",   product: "Widget A", qty: 1, price: 9.99,  shipped: false },
  { customer: "Grace",   product: "Widget C", qty: 3, price: 49.99, shipped: true  },
];

const revenue = orders
  .filter(o => o.shipped)
  .map(o => o.qty * o.price)
  .reduce((sum, v) => sum + v, 0);

const topCustomers = orders
  .filter(o => o.shipped)
  .sort((a, b) => (b.qty * b.price) - (a.qty * a.price))
  .map(o => o.customer)
  .slice(0, 3);

console.log(`Revenue: $${revenue}`);
console.log(`Top customers: ${topCustomers.join(", ")}`);
